### ISSUE TITLE
[JENKINS-62773] Allow for nesting Parallel and Matrix blocks

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/matrix-project-plugin#720]
<p>Apologies if this is a duplicate, but I couldn't find another ticket.</p>

<p>I'd like to create a job that runs a standalone stage, and a matrix based stage, in parallel. Below is an example of what I'm thinking. The AWS job has different steps / params, so it would be run in parallel to the matrix build, which has several more variants.</p>

<p>When running the below, the build stops, and tells me `"matrix" or "parallel" cannot be nested inside another "matrix" or "parallel"`.<br/>
  </p>

<pre>
    stage(<span class="code-quote">'Build Images'</span>) {
      parallel {
stage (<span class="code-quote">'Build AWS'</span>) {
  steps {
    echo <span class="code-quote">"Running AWS Clone Packer Job"</span>
    sh <span class="code-quote">'doing stuff'</span>
  }
}

stage(<span class="code-quote">'Build vCenter'</span>) {
  matrix {
    axes {
      axis {
name <span class="code-quote">'DC'</span>
values <span class="code-quote">'DC1'</span>, <span class="code-quote">'DC2'</span>
      }
      axis {
name <span class="code-quote">'DB'</span>
values <span class="code-quote">'DB1'</span>, <span class="code-quote">'DB2'</span>
      }
      axis {
name <span class="code-quote">'TENNANTS'</span>
values <span class="code-quote">'ST'</span>, <span class="code-quote">'MT'</span>
      }
    }
  }
  steps {
    echo <span class="code-quote">"Running vCenter Clone Packer Job"</span>
    sh <span class="code-quote">"doing something in a matrix"</span>
  }
}
      }
    }
</pre>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=herocc">herocc</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-62773" target="_blank">Allow for nesting Parallel and Matrix blocks</a></i></summary>
<i><ul>
<li><b>status</b>: Open
<li><b>priority</b>: Major
<li><b>component(s)</b>: matrix-project-plugin, pipeline-model-definition-plugin, workflow-cps-plugin
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 17
<li><b>watchers</b>: 17
<li><b>imported</b>: 2025-12-06
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Apologies if this is a duplicate, but I couldn't find another ticket.</p>

<p>I'd like to create a job that runs a standalone stage, and a matrix based stage, in parallel. Below is an example of what I'm thinking. The AWS job has different steps / params, so it would be run in parallel to the matrix build, which has several more variants.</p>

<p>When running the below, the build stops, and tells me `"matrix" or "parallel" cannot be nested inside another "matrix" or "parallel"`.
  </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">    stage(<span class="code-quote">'Build Images'</span>) {
      parallel {
        stage (<span class="code-quote">'Build AWS'</span>) {
          steps {
            echo <span class="code-quote">"Running AWS Clone Packer Job"</span>
            sh <span class="code-quote">'doing stuff'</span>
          }
        }

        stage(<span class="code-quote">'Build vCenter'</span>) {
          matrix {
            axes {
              axis {
                name <span class="code-quote">'DC'</span>
                values <span class="code-quote">'DC1'</span>, <span class="code-quote">'DC2'</span>
              }
              axis {
                name <span class="code-quote">'DB'</span>
                values <span class="code-quote">'DB1'</span>, <span class="code-quote">'DB2'</span>
              }
              axis {
                name <span class="code-quote">'TENNANTS'</span>
                values <span class="code-quote">'ST'</span>, <span class="code-quote">'MT'</span>
              }
            }
          }
          steps {
            echo <span class="code-quote">"Running vCenter Clone Packer Job"</span>
            sh <span class="code-quote">"doing something in a matrix"</span>
          }
        }
      }
    }
</pre>
</div></div></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>CloudBees Core Managed Master 2.204.3.7</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>reddwarf94:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/62773?focusedId=414736&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-414736">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-62773?focusedId=414736&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-414736">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The documentation explicitly mentions this limitation: "It is not possible to nest a <tt>parallel</tt> or <tt>matrix</tt> block within a <tt>stage</tt> directive if that <tt>stage</tt> directive is nested within a <tt>parallel</tt> or <tt>matrix</tt> block itself." (<a href="https://www.jenkins.io/doc/book/pipeline/syntax/#declarative-matrix" class="external-link" target="_blank" rel="nofollow noopener">https://www.jenkins.io/doc/book/pipeline/syntax/#declarative-matrix</a>).</p>

<p>I guess it was left out when adding "matrix" for some good reason. Not that I'm planning to work on this, but maybe it could help somebody else if there were an explanation of the main challenges to implement this (AFAICS <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bitwiseman" class="user-hover" rel="bitwiseman">bitwiseman</a> implemented the matrix support).</p>

<p>People is doing really interesting stuff to overcome this limitation (e.g. <a href="https://stackoverflow.com/questions/63878233/running-non-matrix-stages-in-parallel-to-a-matrix-in-jenkins-declarative-pipelin" class="external-link" target="_blank" rel="nofollow noopener">https://stackoverflow.com/questions/63878233/running-non-matrix-stages-in-parallel-to-a-matrix-in-jenkins-declarative-pipelin</a>).</p>
</pre>
</details>

</ul>
</details>

<p>The documentation explicitly mentions this limitation: "It is not possible to nest a <tt>parallel</tt> or <tt>matrix</tt> block within a <tt>stage</tt> directive if that <tt>stage</tt> directive is nested within a <tt>parallel</tt> or <tt>matrix</tt> block itself." (<a href="https://www.jenkins.io/doc/book/pipeline/syntax/#declarative-matrix" class="external-link" target="_blank" rel="nofollow noopener">https://www.jenkins.io/doc/book/pipeline/syntax/#declarative-matrix</a>).</p>

<p>I guess it was left out when adding "matrix" for some good reason. Not that I'm planning to work on this, but maybe it could help somebody else if there were an explanation of the main challenges to implement this (AFAICS <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bitwiseman" class="user-hover" rel="bitwiseman">bitwiseman</a> implemented the matrix support).</p>

<p>People is doing really interesting stuff to overcome this limitation (e.g. <a href="https://stackoverflow.com/questions/63878233/running-non-matrix-stages-in-parallel-to-a-matrix-in-jenkins-declarative-pipelin" class="external-link" target="_blank" rel="nofollow noopener">https://stackoverflow.com/questions/63878233/running-non-matrix-stages-in-parallel-to-a-matrix-in-jenkins-declarative-pipelin</a>).</p>
